torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
         super().__init__(var_to_sources, lambda n: n.name(), var_to_sources)
 
 
+class DynamicDimConstraintPrinter(StrPrinter):
+    def __init__(self, symbol_to_source):
+        super().__init__()
+        self.symbol_to_source = symbol_to_source
+
+    def print_source(self, source) -> str:
+        return f"dynamic_dim({source.base.name()}, {source.idx})"
+
+    def _print_Symbol(self, expr) -> str:
+        assert isinstance(expr, sympy.Symbol), str(type(expr))
+
+        return self.print_source(self.symbol_to_source[expr][0])
+
+    def _print_Relational(self, expr):
+        return '%s %s %s' % (
+            self.parenthesize(expr.lhs, precedence(expr)),
+            expr.rel_op,
+            self.parenthesize(expr.rhs, precedence(expr))
+        )
+
+
+class DimConstraints:
+    """
+    Custom solver for a system of constraints on symbolic dimensions.
+    Solutions are "static" values or simplified "dynamic" constraints.
+    """
+
+    def __init__(self, symbol_to_source, var_to_val):
+        # We try to solve systems of inequalities with 1 free variable.
+        self._univariate_inequalities = defaultdict(set)
+        # Among them, we prioritize solving for a free variable that has equalities.
+        self._symbols_with_equalities = set()
+        # A solution of a free variable with equalities becomes a substitution.
+        # We use these substitutions to simplify other constraints.
+        self._substitutions = {}
+
+        # In general, constraints may have // and % operations.
+        # Of course, // can be expressed in terms of / and %.
+        # Our inequality solver can handle / but not %. So we need to transform them away.
+        # We do so by using the values of variables as hints to evaluate %.
+        # For soundness we record additional congruence guards and solve them separately.
+        self._var_to_val = var_to_val
+        self._congruences = defaultdict(set)
+
+        # We do not try to (directly) solve inequalities with > 1 free variables.
+        self._multivariate_inequalities = set()
+
+        # We park external equalities between free variables here.
+        self._symbolic_equivalences = []
+
+        # Solutions come in two forms:
+        # - (static) specializations
+        # - (dynamic) inequalities / congruences
+        self._static_results = set()
```

Comment:
what are the types here? I see you throwing strings in?

Reply:
Added types in init.